spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/support/converter/PubSubMessageConverter.java

```diff
@@ -16,58 +16,31 @@
 
 package org.springframework.cloud.gcp.pubsub.support.converter;
 
-import java.io.IOException;
 import java.util.Map;
 
-import com.google.protobuf.ByteString;
 import com.google.pubsub.v1.PubsubMessage;
 
 /**
- * Interface for converters that can convert POJOs to and from PubSub messages.
+ * Interface for converters that can convert POJOs to and from Pub/Sub messages.
  *
  * @author Chengyuan Zhao
+ * @author Mike Eltsufin
  */
 public interface PubSubMessageConverter {
 
 	/**
-	 * Convert a Java object to a byte array payload.
-	 * @param object the object to convert
-	 * @return a byte array ready to be sent with PubSub
-	 */
-	byte[] toPayload(Object object) throws IOException;
-
-	/**
-	 * Convert a payload from a PubSub message into an object of the desired type
-	 * @param payload the {@link ByteString} payload from the message
-	 * @param payloadType the desired type of the object
-	 * @return the object converted from the payload of the message
-	 */
-	<T> T fromPayload(ByteString payload, Class<T> payloadType) throws IOException;
```

Comment:
Why get rid of these two methods? By removing these two and the default implementations of toMessage and fromMessage you've now repeated the code for dealing with the headers and also message.getData() calls in both the jackson and simple message converter classes.

Reply:
I think it's better handled by a utility class if we need this.
I feel it's best to keep the interface as minimal as possible. There are cases where you would want to override `toMessage` and `fromMessage`, and it would be additional burden for those implementation to also implement `toPayload` and `fromPayload`.

Reply:
Could you elaborate? I don't follow. Anybody implementing this interface necessarily implements the functionality of both To/From message and To/From Payload. The difference that we're discussing is whether that is split into 2 methods or 4 methods.  In fact, reusing 2 out of those 4 methods reduces burden. Currently it looks like Jackson and Simple both repeat what was in the to/from Message methods' default implementations, so that actually is the additional burden.

Reply:
If you imagine a converter that relies on headers to make the decision, for example, one that uses something like a mime-type header, 2 of the 4 methods would just become a burden. As a best practice, interfaces should always be kept as minimal as possible, preferably even limited to a single method to be compatible with lambdas.

Reply:
Well, I'd prefer to keep only two, message-based methods. Even if our out-of-the-box implementations have some similar code which may go to the `default` methods, there is no guarantee that some other implementations won't have fully different logic, so end-user will have to implement 4 methods instead of 2, where those `to(from)Payload()` may be just with empty bodies because they are not going to be called from others.

Plus with 4 methods there can be some confuse which method to choose to use in the API consumer.